Adding basic functionality to work with google maps

dezd4r opened the pull request:
Есть несколько проблем: 

1. Если приложению не дать разрешения на изпользование геолокации или дать, но навсегда, оно перестанет запускаться или вовсе вылетит, надо будет включать GPS после разрешения на его использование(это не сделано).
2. Зона действия игрока отрисовывается единожды, но это решается отдельным поток, который раз  внесколько секунду будет просить координаты игрока и по ним все перерисовывать.

dezd4r committed:
Initial commit

dezd4r committed:
Initial commit

dezd4r committed:
add google maps

dezd4r committed:
add google maps

dezd4r committed:
add style for map and app

dezd4r committed:
add style for map and app

dezd4r committed:
add location and activity zone of user

dezd4r committed:
add location and activity zone of user

dmitrii-artuhov commented:
Из замечаний не относящихся к коду:
- наверное, стоит заменить стиль для карты, так как на этой не видно подписей населенных пунктов, так как текст сливается с задним фоном (можно [этот](https://snazzymaps.com/style/50689/old-time-fantasy) попробовать) 
- на моем телефоне не появлялась кнопка геолокации (то есть я не мог свою позицию посмотреть), пока не переоткрыл приложение
- при запуске на пк эмуляторе я находился где-то на западном побережье США, неверно геолокация сработала, хотя впн не был включен. Мб, я на пк как-то неверно запускаю его и надо по-другому права на получение геолокации дать?
- про проблему 1, что ты описал выше: даже если геолокация включена, приложение показывает `Toast` со словами включите геолокацию (забавно, что если после этого переоткрыть приложение, то снова все окей)

dmitrii-artuhov commented on app/src/main/java/com/gigachadllc/enchanted_towers/MapsActivity.java:
```diff
@@ -0,0 +1,120 @@
+
+        LocationManager locationManager = (LocationManager) getSystemService(Context.LOCATION_SERVICE);
+
+        //Toast.makeText(this, "Please turn on GPS", Toast.LENGTH_SHORT).show();
```
если комментарии не нужны, то лучше их убрать совсем

dmitrii-artuhov commented on app/src/main/java/com/gigachadllc/enchanted_towers/MapsActivity.java:
```diff
@@ -0,0 +1,120 @@
+
+import android.widget.Toast;
+
+public class MapsActivity extends AppCompatActivity
```
лучше везде переименовать на `MapActivity`, в том числе и имена файлов

vvselishchev replied:
+1

dmitrii-artuhov commented on app/src/main/java/com/gigachadllc/enchanted_towers/MapsActivity.java:
```diff
@@ -0,0 +1,120 @@
+
+        if (ContextCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION)
+                == PackageManager.PERMISSION_GRANTED /*|| ContextCompat.checkSelfPermission(this, Manifest.permission.ACCESS_COARSE_LOCATION)
+                == PackageManager.PERMISSION_GRANTED*/) {
```
аналогично про комментарии

dmitrii-artuhov commented on app/src/main/java/com/gigachadllc/enchanted_towers/MapsActivity.java:
```diff
@@ -0,0 +1,120 @@
+            Toast.makeText(this, "Please turn on GPS", Toast.LENGTH_SHORT).show();
+            return;
+        }
+        //Toast.makeText(this, "Please turn on GPS", Toast.LENGTH_SHORT).show();
```
аналогично про комментарии

dmitrii-artuhov commented on app/src/main/java/com/gigachadllc/enchanted_towers/MapsActivity.java:
```diff
@@ -0,0 +1,120 @@
+        ActivityCompat.OnRequestPermissionsResultCallback {
+
+    private static final int LOCATION_PERMISSION_REQUEST_CODE = 1;
+    Location location;
```
`private Location location;`

dmitrii-artuhov commented on app/src/main/java/com/gigachadllc/enchanted_towers/MapsActivity.java:
```diff
@@ -0,0 +1,120 @@
+
+        //Toast.makeText(this, "Please turn on GPS", Toast.LENGTH_SHORT).show();
+
+        if (ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED && ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
```
вот эти проверки `checkSelfPermission(...) == ...` часто повторяются. Лучше вынести в отдельный метод и как-то унифицировать.

dmitrii-artuhov commented on app/src/main/java/com/gigachadllc/enchanted_towers/MapsActivity.java:
```diff
@@ -0,0 +1,120 @@
+
+    private void enableMyLocation() {
+
+        if (ContextCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION)
```
то же самое - про повторение кода для проверки прав доступа

dezd4r committed:
bug fix

dezd4r committed:
bug fix

dezd4r committed:
style change

dezd4r committed:
style change

vvselishchev commented on app/src/main/java/com/gigachadllc/enchanted_towers/MapsActivity.java:
```diff
@@ -0,0 +1,149 @@
+    }
+
+    @Override
+    public void onMapReady(@NonNull GoogleMap googleMap) {
```
Длинный метод, надо разбить на маленькие логические куски.

vvselishchev commented on app/src/main/java/com/gigachadllc/enchanted_towers/MapsActivity.java:
```diff
@@ -0,0 +1,149 @@
+
+        SupportMapFragment mapFragment =
+                (SupportMapFragment) getSupportFragmentManager().findFragmentById(R.id.map);
+        assert mapFragment != null;
```
Лучше выкинуть ошибку, ассерты в продакшене не используются

vvselishchev commented on app/src/main/java/com/gigachadllc/enchanted_towers/MapsActivity.java:
```diff
@@ -0,0 +1,149 @@
+            builder.setNegativeButton("Cancel", new DialogInterface.OnClickListener() {
+                @Override
+                public void onClick(DialogInterface dialog, int which) {
+                    //Nothing
```
А зачем её оверрайдить тогда?

vvselishchev commented on app/src/main/java/com/gigachadllc/enchanted_towers/MapsActivity.java:
```diff
@@ -0,0 +1,149 @@
+            }
+        };
+
+        locationManager.requestLocationUpdates(LocationManager.GPS_PROVIDER, 0, 0, locationListener);
```
Что такое 0, 0? Лучше оставить комментарий или сделать константу

vvselishchev commented on app/src/main/java/com/gigachadllc/enchanted_towers/MapsActivity.java:
```diff
@@ -0,0 +1,149 @@
+    }
+
+    private void drawCircle(LatLng point){
+        CircleOptions circleOptions = new CircleOptions();
```
Этот объект всегда константный? Если да, то, может, лучше не создавать каждый раз, а сделать полем и инициализировать там же, а тут просто его подвинуть. CircleOptions реализует шаблон Builder, поэтому можно писать просто:
```
new CircleOptions()
  .strokeColor(radius)
  .fillColor(...)
  .strokeWidth(...)
```

vvselishchev commented on app/src/main/java/com/gigachadllc/enchanted_towers/MapsActivity.java:
```diff
@@ -0,0 +1,149 @@
+        map.setOnMyLocationButtonClickListener(this);
+        map.setOnMyLocationClickListener(this);
+        while (!map.isMyLocationEnabled()) {
+            enableMyLocation();
```
Просто интересно, что будет с этим циклом, если разрешение долго не дают? Есть ли внутри какой-то sleep между итерациями? Что будет, если вывести через Toast что-нибудь внутри цикла?

Vladislav0Art replied:
Возможно, запрос прав правильнее делать через event'ы, чтобы таким aka `while(true)` (если `enableMyLocation` неблокирующий) не убивать CPU. 

Особо сильно не вникал, как это правильно делать, но у `ActivityCompat` есть interface [`ActivityCompat.OnRequestPermissionsResultCallback`](https://developer.android.com/reference/androidx/core/app/ActivityCompat.OnRequestPermissionsResultCallback), от которого можно отнаследоваться и при получении доступа прав создавать нужный activity.

У Android есть репозиторий с тем, как это правильно делать: [permissions-samples](https://github.com/android/permissions-samples). Наш случай, скорее всего [permissions-samples/RuntimePermissionsBasic](https://github.com/android/permissions-samples/tree/c4573f1218cef81295b519592718ecef3b144095/RuntimePermissionsBasic). 

Если в данном примере посмотреть в [`MainActivity.java`](https://github.com/android/permissions-samples/blob/c4573f1218cef81295b519592718ecef3b144095/RuntimePermissionsBasic/Application/src/main/java/com/example/android/basicpermissions/MainActivity.java#L100), то у него, как я понял, такой workflow:

1. Создать кнопку для показа камеры.
2. При нажатии, если права уже есть, то показать камеру, иначе - сделать запрос прав (как раз через `ActivityCompat.requestPermissions(Activity, String[], int)`).
3. Результат запроса на получение прав прилетает в вызове метода `ActivityCompat.onRequestPermissionsResult`, в котором запускается activity камеры, если права были получены, иначе ничего не делается.

В самом файле есть более детальное описание. В эту сторону, думаю, нужно посмотреть.

Vladislav0Art replied:
Данный `while` будет постоянно выполнять код своего тела (попробовал на `println`). Решаем проблему.

Vladislav0Art commented:
Нужно текущие изменения в папке `app/` переложить в папку `client/app` (`app/build.gradle` нужно поместить либо в `client/app/build.gradle`, либо в `build-logic/android-application/build.gradle` или в папку `build-logic/android-application/src` - там есть gradle-файл).

dezd4r committed:
Delete redundant project root build.gradle

dezd4r committed:
Add depandency on play-services-maps into client:app module

Vladislav0Art commented on app/src/main/java/com/gigachadllc/enchanted_towers/MapsActivity.java:
```diff
@@ -0,0 +1,149 @@
+        circleOptions.center(point);
+        circleOptions.radius(200);
+        circleOptions.strokeColor(Color.BLACK);
+        circleOptions.fillColor(0x30ff0000);
```
Лучше в константу вынести, название которой бы отражало этот цвет. Например, `TRANSLUCENT_RED = 0x30ff0000`.

vvselischev replied:
Мне кажется, это уже должно быть реализовано где-то. Конвертер из enum-а цветов в 16-ричное представление (но не уверен).

dezd4r committed:
Try add into master

Vladislav0Art committed:
Merge branch 'vnaumkin/feature/add-google-maps' of https://github.com/Giga-Chad-LLC/Enchanted-Towers into vnaumkin/feature/add-google-maps

Vladislav0Art committed:
Remove redundant root build.gradle file

Vladislav0Art committed:
Remove obsolete/commented code from MainActivity & MapActivity

Vladislav0Art committed:
Remove obsolete layout elements from activity_main.xml

Remove via Android Studio Editor, thus the xml file is valid.

Vladislav0Art committed:
Rename previous fragment_map.xml file (mark it as old)

Vladislav0Art committed:
Create blank fragment (boilerplate for MapFragment)

Vladislav0Art committed:
Create MapFragment that supports Google Map

MapFragment java class corresponds to fragment_map.xml file.

Vladislav0Art committed:
Mount MapFragment into MapActivity

Vladislav0Art committed:
Refactor code of MapFragment class

Vladislav0Art committed:
Add default layout into fragment_map.xml (tools:layout)

Vladislav0Art committed:
Replace map fragment id with correct one (google_map_fragment)

Vladislav0Art committed:
Add custom map style & add logging into MapFragment

Vladislav0Art committed:
Modify activity_map & fragment_map xml files

Vladislav0Art committed:
Move location related logic into MapFragment class

Vladislav0Art committed:
Remove commented code from MapActivity class

Vladislav0Art committed:
Add links to important resource about runtime permissions

Vladislav0Art committed:
Remove commented obsolete code from MainActivity

Vladislav0Art committed:
Create core workflow of accessing location permissions in MapActivity

Vladislav0Art committed:
Mount map fragment if location permissions granted

Vladislav0Art committed:
Establish correct workflow of MapFragment

Now MapFragment checks whether both of the required permissions granted (these are ACCESS_FINE_LOCATION & ACCESS_COARSE_LOCATION)
as well as checks the GPS Provider availability, and asks to enable it if disabled.

The mentioned checks make the MapFragment component resistant to accidental GPS signal losses and manual permissions denial (via app settings of smartphone).

Vladislav0Art committed:
Move MapFragment logic into separate methods

Moved map style applying & registering of click event listeners into separate methods of MapFragment class.

Vladislav0Art committed:
Move permissions checks in MapFragment into method

Name of the method must satisfy certain name constraints (special case of Android Studio).
See: [Check Android Permissions in a Method](https://stackoverflow.com/questions/36031218/check-android-permissions-in-a-method)

Vladislav0Art committed:
Move registration of location updates listener into separate method

Vladislav0Art committed:
Draw circle around last known location in MapFragment

Vladislav0Art committed:
Make integer literals in requestLocationUpdates call be named constants

Vladislav0Art committed:
Create single AlertDialog component & unregister updates listener on destroy event

AlertDialog should be singleton (not to create it each time user turns off GPS).
Unregister updates listener to stop acting when the MapActivity is destroyed (i.e. closed).

Vladislav0Art committed:
Add TODO about MapFragment's fields being null

Vladislav0Art committed:
Move onDestroy method to the end of file

Vladislav0Art committed:
Create PermissionManager & create static method checkLocationPermission

Vladislav0Art committed:
Add null check into MapFragment in onDestroy hook

It was causing crush of the app when allowing/denying device location access manually from settings.

Vladislav0Art committed:
Create hook-methods to conform proper permission workflow in PermissionManager

Vladislav0Art committed:
Create permission workflow with UI in MapActivity

Currenlty we have the following flow:
1. If all location permission granted -> mount google map.
2. If any permission absents, either:
    1. Show rationale with info why permissions needed:
        1. If user accepts -> show permissions grant panel.
        2. If user rejects -> limit content (do not mount google map).
    2. Directly show permissions grant panel.
3. Resulting permissions:
    1. At least coarse location permission granted -> mount google map.
    2. No location permissions -> limit content (do not mount google map).

Vladislav0Art commented:
В последнем коммите указал, какой сейчас workflow по получению прав к геолокации и показу карты. Дублирую тут:

### Currenlty we have the following flow:
1. If **all** location permissions already granted -> mount google map.
2. If **any** permission absents, **either**:
    1. Show rationale with info why permissions needed (_this applies depending on OS version_):
        1. If user **accepts** dialog -> show permissions grant panel.
        2. If user **rejects** dialog  -> limit content (do not mount google map).
    2. Directly show permissions grant panel.
3. **Resulting permissions:**
    1. **At least** coarse location permission granted -> mount google map.
    2. **No** location permissions -> limit content (do not mount google map).

В случае, если пользователь дал права только на `coarse location`, то при каждом заходе мы будет просить права на точное местоположение (`fine location`), т.к. `coarse location` довольно [неточный](https://developer.android.com/training/location/permissions#accuracy).

Vladislav0Art committed:
Make circleOptions member of MapFragment class

Vladislav0Art committed:
Set fillColor prop of circleOptions with Color.argb method

dmitrii-artuhov commented on client/app/src/main/java/enchantedtowers/client/MainActivity.java:
```diff
@@ -18,128 +19,22 @@
-
 
 public class MainActivity extends AppCompatActivity {
+    //Fragment fragment = new MapFragment();
```
Если комментарий не нужен, то удалить

Vladislav0Art replied:
done

dmitrii-artuhov commented on client/app/src/main/java/enchantedtowers/client/MapActivity.java:
```diff
@@ -0,0 +1,131 @@
+
+        ActivityResultLauncher<String[]> locationPermissionLauncher =
+                registerForActivityResult(new ActivityResultContracts.RequestMultiplePermissions(), result -> {
+                    Boolean fineLocationPermissionGranted = result.getOrDefault(
```
почему просто на написать в этом месте `boolean` вместо обертки?

Vladislav0Art replied:
Добавил TODO

dmitrii-artuhov commented on client/app/src/main/java/enchantedtowers/client/MapActivity.java:
```diff
@@ -0,0 +1,131 @@
+                    }
+                });
+
+        /*if (PermissionManager.checkLocationPermission(this)) {
```
Удалить комментарий, если не нужен

dmitrii-artuhov commented on client/app/src/main/java/enchantedtowers/client/MapFragment.java:
```diff
@@ -0,0 +1,227 @@
+
+
+public class MapFragment extends Fragment {
+    // TODO: make the following fields Optional<T>
```
Надо сделать

Vladislav0Art replied:
Потом

Vladislav0Art committed:
Delete obsolete app/ folder since its content now in client

dmitrii-artuhov commented on client/app/src/main/res/layout/activity_map.xml:
```diff
@@ -0,0 +1,25 @@
+    tools:context=".MapActivity">
+
+
+    <!--    <com.google.android.gms.maps.MapView-->
```
Удалить комментарий

Vladislav0Art committed:
Remove redundant comment

Vladislav0Art committed:
Remove obsolete commented code

Vladislav0Art committed:
Remove obsolete commented layout in activity_map.xml

Vladislav0Art committed:
Merge branch master into vnaumkin/feature/add-google-maps
